getdns_and_stubby: Initial commit for lede/openwrt support

iamperson347 opened the pull request:
Signed-off-by: David Mora <iamperson347+public@gmail.com>

Maintainer: me / @iamperson347
Compile tested: ar71xx, archer-c7, trunk
Run tested: same as above. Tested on personal Archer C7 v2. Functionality seems to be correct.

Description: Initial commit for support of using getdns library and stubby deamon on lede/openwrt firmware.

https://getdnsapi.net/
https://dnsprivacy.org/wiki/display/DP/DNS+Privacy+Daemon+-+Stubby

karlp commented:
The depends should be in the package makefile, you only need config files if you _realllly_ need them.

You shouldn't have a copyright statement for openwrt from 2010, the copyright statement should either be your name, or omitted entirely.  (Unless you actually have a contract with openwrt.org)

Is the GETDNS_ENABLE_STUB_ONLY option actually useful?  Would anyone actually run it like that?  Try not to make endless config options, the binary packages built can only have the defaults.

Package/getdns/description says it contains the library, but you also have libgetdns?  Perhaps you could clarify the desriptions, if not hing else

You only have the one +$(eval $(call BuildPackage,libgetdns))  line, so it seems jsut obfuscated to try having both "getdns" and "libgetdns" ?

stubby has the same comments on copyright

stubby has duplicate descriptions.

stubby's conffile declaration looks wrong, refers to /usr/local, but the install is to /etc? 

stubby and getdns should be in one commit each, with better commit titles.

your stubby init script shouldn't really copy all the text from wiki page on procd scripts :)
Do you actually need pidfiles?  (you normally don't need these unless you're working with a different monitoring system)

Sorry I couldn't put all this inline, github review web wasn't supporting my browser properly :(

iamperson347 commented:
Hi Karlp, thanks for all of the feedback. If you couldn't tell, this is my first attempt to make a package for opewrt. I'll work through these items (when I get a chance) and get them checked in/provide additional feedback.

WRT having separate commits, should I make two separate pull requests? Or should I just revert current commit and check in two new commits with feedback implemented?

iamperson347 commented:
(Keep in mind, I'm not the developer of getdns, but I will answer the best I can.)

You shouldn't have a copyright statement for openwrt from 2010, the copyright statement should either be your name, or omitted entirely. (Unless you actually have a contract with openwrt.org)
- Fixed

Is the GETDNS_ENABLE_STUB_ONLY option actually useful? Would anyone actually run it like that? Try not to make endless config options, the binary packages built can only have the defaults.
- For use on hardware running openwrt, I think this would be the default. I threw in the config option just in case something down the line would like to use more functionality of the getdns library. For normal use cases, my assumption is that people would use one of the regular recursive dns resolvers (unbound), and then forward to stubby, which will in turn encrypt DNS queries via TLS.
- Enabling GETDNS_ENABLE_STUB_ONLY (configure option --enable-stub-only) also removes the libunbound dependency (make sense if someone wanted to use something like dnsmasq instead of unbound).

Package/getdns/description says it contains the library, but you also have libgetdns? Perhaps you could clarify the desriptions, if not hing else
- Fixed, this was remnants of me originally packing both the library and daemon together (since stubby can be compiled as a submodule of getdns). 

You only have the one +$(eval $(call BuildPackage,libgetdns)) line, so it seems jsut obfuscated to try having both "getdns" and "libgetdns" ?
- Fixed, see above

stubby has the same comments on copyright
- Fixed

stubby has duplicate descriptions.
- Fixed

stubby's conffile declaration looks wrong, refers to /usr/local, but the install is to /etc?
- Fixed

stubby and getdns should be in one commit each, with better commit titles.
- Fixed (I think)

your stubby init script shouldn't really copy all the text from wiki page on procd scripts :)
Do you actually need pidfiles? (you normally don't need these unless you're working with a different monitoring system)
- Fixed and tested. I don't believe stubby explicitly needs a pid file, and procd appears to handling it correctly (since stubby can be run in the foreground, I assume this is how procd keeps track?)

Thanks for working with me on this. Please let me know your thoughts.

karlp commented:
Looks pretty good to me now.  To cleann up the commits, just rebase -i and drop out the first two commits.  You should end up with just the later two commits.  If you then "push -f" to the same branch name, the github PR updates automatically.

karlp commented:
You'll still need to deal with hthe correct Signed-oFF-by lines for travis of course :)

iamperson347 commented:
karlp - Thanks for the help so far.

I think I have the commits straightened out. Travis is now complaining that the download step has failed on both packages (WARNING or ERROR matched). However, I'm not seeing any issues on my side when I run 'download' or 'check' on both packages. Download URLs are good, sha256 hashes should match, etc.

Is there anything you can think of that would be causing this?

karlp commented:
Not sure on that one sorry.  I know there's some problems with the unnamed download files though sometimes.  perhaps try the form

PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
PKG_SOURCE_URL:=https://github.com/user/project/archive/$(PKG_VERSION)

instead of the just PKG_SOURCE=v$(PKG_VERSION).tar.gz style?  It makes sure that the downloaded file has a name more likely to be properly unnique, though that shouldn't be a problem for travis?

iamperson347 commented:
Seems like some other pull requests are having issues with the download check as well (curl finishes the download but then the check fails). One of the proposed packages on this PR (getdns) follows the {package}-{version} scheme (unique name) but fails as well (with Travis).

champtar commented:
#5448

iamperson347 commented:
Thanks @champtar - the changes to the travis checks allowed the builds to pass.

Is there anything else I need to do with this PR?

karlp commented on net/stubby/Makefile:
```diff
@@ -0,0 +1,59 @@
+	CATEGORY:=Network
+	SUBMENU:=IP Addresses and Names
+	TITLE+= - (daemon that uses getdns)
+	USERID:=stubby=891:stubby=891
```
what's the latest wisdom on USERID:=stubby=891:stubby=891 type things? did we ever 
               get to an autoassignment of them? or a registry?

iamperson347 replied:
I definitely just picked the uid/gid randomly. The upstream source is using "stubby" as the username and group name.

 A registry sounds like a great idea to avoid conflicts.

karlp replied:
right, I was just checking elsewhere, this is as good as you can do right now I think.

champtar replied:
At what number does user start on OpenWrt? (on some linux first user is uid 500)

karlp commented:
:+1:

poranje commented:
There was a thread on "[LEDE-DEV] convention on uid/gid for packages" that ran in May 2017.

> Op 19 jan. 2018, om 13:12 heeft champtar <notifications@github.com> het volgende geschreven:
> 
> @champtar commented on this pull request.
> 
> In net/stubby/Makefile:
> 
> > +define Package/stubby/Default
> +	TITLE:=stubby
> +	URL:=https://dnsprivacy.org/wiki/display/DP/DNS+Privacy+Daemon+-+Stubby
> +endef
> +
> +define Package/stubby/description
> +	This package contains the Stubby daemon (which utilizes the getdns library).
> +endef
> +
> +define Package/stubby
> +	$(call Package/stubby/Default)
> +	SECTION:=net
> +	CATEGORY:=Network
> +	SUBMENU:=IP Addresses and Names
> +	TITLE+= - (daemon that uses getdns)
> +	USERID:=stubby=891:stubby=891
> 
> At what number does user start on OpenWrt? (on some linux first user is uid 500)
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

iamperson347 commented:
Checking in on the status of this PR. I checked the mail thread that was mentioned with regards to uid/gid, and I didn't feel there was a definitive answer/best practice yet. If reviewers desire, I can test changing to the "auto assign" method.

champtar commented:
@iamperson347 just to be sure can you pick an uid/gid <500
@karlp as you've done the initial review i'll let you merge ;)

iamperson347 commented:
Hey everyone - I have adjusted the uid/gid. I didn't see any conflicts when I did a search through the package repo.

karlp commented:
merged in bbab2838 and 3e56158e thanks
